sdk/helper/storagepacker/storagepacker_v2.go

```diff
@@ -81,6 +81,156 @@ type LockedBucket struct {
 	*Bucket
 }
 
+type lockOperation func(*LockedBucket)
+
+// Safely identify the bucket for a given key, and aquire its storage lock
+// The key may be the full hash of the item ID, or a prefix, with a bucket Key with embedded /'s
+// The cache may initially be empty, so "not found" is acceptable, and needs to be handled
+// by deciding whether to create the bucket.
+//
+// This function uses the same logic whether we want a read or a write operation.
+// Use the helpers below that set everything up nicely for read locks or write locks.
+func (s *StoragePackerV2) lockBucket(key string, acquire lockOperation, release lockOperation) (bucket *LockedBucket, found bool, err error) {
```

Comment:
I don't really understand why we need this function. All buckets are loaded into the cache on startup. We should be able to simply find the bucket in the cache and use its lock directly.

Reply:
That's the race condition I identified.

If we're doing a GetItem() or PutItem(), we need to identify which bucket to lock. After a thread calls LongestPrefix and calls bucketsCacheLock.RUnlock(), there's a window where that bucket could be sharded, so it's no longer the right bucket.  The code here acquires the bucket lock and then checks that it's still the right bucket.

The alternative approach of acquiring the bucket lock while we still hold the bucketsCacheLock leads to deadlock, when we update the bucketsCacheLock during sharding.  (Unless we release the BucketLock during the sharding, which I don't think can be made to work reasonably.)

Reply:
I see. I thought that in the new paradigm when we were sharding we were holding a write lock on bucketsCacheLock during the process.

Reply:
That still wouldn't eliminate the race condition as long as there is a gap between releasing a read lock on bucketsCacheLock and acquiring a read or write bucketLock.  We could of course turn bucketsCacheLock into a global lock for all operations.

We could hold the bucketsCacheLock for the duration of sharding, and also reverse the lock order so that bucket locks could be acquired while bucketsCacheLock was held. That's blocks everybody for the duration of the 17 write operations, but I don't think it's all that much simpler since in the write path you still have to re-acquire the bucketsCacheLock when sharding, even though you dropped it for most single-item writes.